sklearn/preprocessing/data.py

```diff
@@ -2089,11 +2090,32 @@ def fit(self, X, y=None):
             raise ValueError('QuantileTransformer only accepts non-negative'
                              ' sparse matrices')
 
+        # check the number of quantiles is less than the number of samples used
+        # in fitting
+        if (self.n_quantiles > min(X.shape[0], self.subsample)):
+            raise ValueError('The number of quantiles is less than either the'
+                             ' number which will be used during fitting. Got'
+                             ' {} quantiles for {} samples.'.format(
+                                 self.n_quantiles, min(X.shape[0],
+                                                       self.subsample)))
+
         if sparse.issparse(X):
             self._sparse_fit(X)
         else:
             self._dense_fit(X)
 
+        # check that the quantiles are strictly monotonically increasing
+        for feature_idx, quantile in enumerate(self.quantiles_):
+            if not np.all(np.diff(quantile) > 0):
+                warnings.warn('The feature values corresponding to the'
+                              ' quantiles computed are not strictly'
+                              ' monotonically increasing for the feature #'
+                              '{}. This configuration is ill-posed for the'
```

Comment:
This is wrong. The feature values are not required to be strictly monotonically increasing. Samples are not supposed to be in any particular order.

Let me suggest:

```
The computed quantiles for feature %d are not strictly monotonically increasing. This configuration is ill-posed and can lead to invalid transformation. If this is feature is actually a numerical feature, reducing the number of requested quantiles of increasing the size of the training set or subsample might fix the issue. This can also happen if a discrete value is repeated many times in the data in which case adding some very small noise will fix the issue. Note that QuantileTransfomer should never be applied to categorical features.
```